design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,270 @@
+
+#### Scenario 3: Infrastructure owner enables self-service, except for one on-prem cluster with sensitive data
+
+1. Infrastructure owner creates a variety of Kubernetes cluster classes with
+   configuration for different environments and different cloud providers.
+   Application owners are given permissions to create clusters using these
+   classes as they see fit.
+1. There is one particular set of workloads that is required to be run in an
+   existing on-premises cluster (for some important reason). The infrastructure
+   owner creates a Kubernetes `Secret` with connection information to that
+   cluster in an infrastructure namespace (secrets are namespace-scoped). They
+   then create a cluster-scoped Kubernetes `Provider` that references the
+   `Secret`.
+1. Application owner says that they need to run a sensitive workload on that
+   cluster. Infrastructure owner creates a new namespace called
+   `super-secret-workloads` and creates a `WorkloadTarget` in that namespace
+   that references the previously created Kubernetes `Provider`.
+1. Application owner provisions a `KubernetesApplication` with labels that
+   select that `WorkloadTarget` for provisioning.
+
+This example shows the power of this model, but also one of the notable
+shortcomings: if a user can publish clusters to a namespace with a
+`WorkloadTarget`, then they can consume any cluster. For the application owners
+that are doing self-service provisioning in this scenario, they will likely need
+to be able to create `WorkloadTarget` resources for their dynamically
+provisioned Kubernetes cluster managed resources. However, if they can do so,
+they could just as easily create a `WorkloadTarget` for the Kubernetes
+`Provider` that is only to be used for sensitive workloads. While it would be
+nice to restrict the clusters that could be published by users with the ability
+to create `WorkloadTarget` resources, there is not an immediately clear and
+elegant way to do so. It could be argued that only trusted users should have
+this permission level anyway, so it may be less of an issue than expressed here.
+Nevertheless, a potential solution is outlined [below](#future-considerations).
+
+## Technical Implementation
+
+The implementation of this proposal is relatively lightweight and could
+potentially be implemented without churn on any existing exposed APIs. The first
+step would be implementing the `WorkloadTarget` type in [core Crossplane]. It
+would not require any additional controllers.
+
+The next step would be to modify the `KubernetesApplication` [scheduling
+controller] to list `WorkloadTarget` resources in the namespace that match
+labels, instead of the current implementation that lists `KubernetesCluster`
+claims. Currently, this controller sets a reference to a `KubernetesCluster`
+claim, which is then later propagated to the child
+`KubernetesApplicationResource` objects, which use it to get a `Secret` with
+[remote cluster connection information].
```

Comment:
It may make sense to rename the `KubernetesApplication` `clusterRef` field if we're scheduling to something other than `KubernetesCluster`. Maybe `targetRef`? I don't think renaming that field would be too much of an issue given it's relatively rarely explicitly set in current examples, and is a `v1alpha1` API.

Reply:
@negz that sounds good to me! @muvaf I definitely can understand that desired experience, but one of the major values of the design here is that we are isolating the *publishing* of k8s clusters with a `WorkloadTarget`. If a cluster could be referenced directly, then any cluster in the Crossplane cluster could be used. To improve this user experience when using self-service cluster provisioning, I have proposed a solution in the `Future Considerations` section.

Reply:
> If a cluster could be referenced directly, then any cluster in the Crossplane cluster could be used.

Not any but the ones in your app namespace. I don't know, as a user, having to create another resource while I have `KubernetesCluster` in my namespace feels odd to me.

Even though we do the side-effect `WorkloadTarget` creation, I'd still want to be able to refer to what I created directly. Thinking of as a stack developer or helm chart author of my app, this would lower the burden of having to realize those relations in the provisioning logic.